requestparsing/test_harness.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2020 Google LLC
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package requestparsing
+
+import (
+	"context"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+	"sync"
+)
+
+// AssertHandler is used to assert properties about the http.Request that it receives in using a callback function.
+type AssertHandler struct {
+	callback func(*http.Request)
+}
+
+func (h *AssertHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	h.callback(r)
+	io.WriteString(w, "Hello world!")
+}
+
+type fakeListener struct {
+	closeOnce      sync.Once
+	channel        chan net.Conn
+	serverEndpoint io.Closer
+	clientEndpoint net.Conn
+}
+
+// newFakeListener creates an instance of fakeListener. This will pass requests to the HTTP server as part of the testing harness.
+func newFakeListener() *fakeListener {
+	s2c, c2s := net.Pipe()
+	c := make(chan net.Conn, 1)
+	c <- s2c
+	return &fakeListener{
+		channel:        c,
+		serverEndpoint: s2c,
+		clientEndpoint: c2s,
+	}
+}
+
+// Accept passes a network connection to the HTTP server to enable bidirectional communication with the client.
+// It will return an error if Accept is called after the listener was closed.
+func (l *fakeListener) Accept() (net.Conn, error) {
+	ch, ok := <-l.channel
+	if !ok {
+		return nil, errors.New("Listener closed")
+	}
+	return ch, nil
+}
+
+// Close will close the two network connections and the listener.
+func (l *fakeListener) Close() error {
+	l.closeOnce.Do(func() {
+		close(l.channel)
+	})
+	err := l.serverEndpoint.Close()
+	err2 := l.clientEndpoint.Close()
+	if err2 != nil {
+		return err2
+	}
+	return err
+}
+
+// Addr returns the network address of the client endpoint.
+func (l *fakeListener) Addr() net.Addr {
+	return l.clientEndpoint.LocalAddr()
+}
+
+// SendRequest writes a request to the client endpoint connection. This will be passed to the server through the listener.
+// The function blocks until the server has finished reading the message.
+func (l *fakeListener) sendRequest(request []byte) error {
+	n, err := l.clientEndpoint.Write(request)
```

Comment:
Check the error first, then the length.

Usually, if the returned error is not `nil`, you should not assume anything about the other values.

Reply:
Resolved in [527b6aa](url)